slog.go

```diff
@@ -0,0 +1,81 @@
+//go:build go1.21
+// +build go1.21
+
+/*
+Copyright 2023 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logr
+
+import (
+	"log/slog"
+)
+
+// SlogImplementor is an interface that a logr.LogSink can implement
+// to support efficient logging through the slog.Logger API.
+type SlogImplementor interface {
+	// GetSlogHandler returns a handler which uses the same settings as the logr.LogSink.
+	GetSlogHandler() slog.Handler
+}
+
+// LogrImplementor is an interface that a slog.Handler can implement
+// to support efficient logging through the logr.Logger API.
+type LogrImplementor interface {
+	// GetLogrLogSink returns a sink which uses the same settings as the slog.Handler.
+	GetLogrLogSink() LogSink
+}
+
+// ToSlog returns a slog.Logger which writes to the same backend as the logr.Logger.
+func ToSlog(logger Logger) *slog.Logger {
+	return slog.New(ToSlogHandler(logger))
+}
+
+// ToSlog returns a slog.Handler which writes to the same backend as the logr.Logger.
+func ToSlogHandler(logger Logger) slog.Handler {
+	if slogImplementor, ok := logger.GetSink().(SlogImplementor); ok {
+		handler := slogImplementor.GetSlogHandler()
+		return handler
+	}
+
+	return &slogHandler{sink: logger.GetSink()}
+}
+
+// FromSlog returns a logr.Logger which writes to the same backend as the slog.Logger.
+func FromSlog(logger *slog.Logger) Logger {
+	return FromSlogHandler(logger.Handler())
+}
+
+// FromSlog returns a logr.Logger which writes to the same backend as the slog.Handler.
+func FromSlogHandler(handler slog.Handler) Logger {
+	if logrImplementor, ok := handler.(LogrImplementor); ok {
+		logSink := logrImplementor.GetLogrLogSink()
+		return New(logSink)
+	}
+
+	return New(&slogSink{handler: handler})
+}
```

Comment:
It would be good if `FromSlog(ToSlog(...))` and vice versa returned the original logger. This currently works when the backend supports both APIs. But if not, we end up layering slogSink on top of slogHandler or vice versa.

Can be fixed with the "underlier" pattern.

Reply:
I skipped over one detail: `logr.Logger.level` gets lost during conversion.

Suppose someone does this:
```
func foo(logger logr.Logger) {
    // The Info call could be in some other library which only uses slog.
    logr.ToSlog(logger).Info("hello")
}

func bar(logger logr.Logger) {
  foo(logger.V(4))
}
```

Should that info message then become a debug message (info - 4 = debug in slog)? I think the answer is yes.

This can be done in the `slogHandler` or by wrapping a real `slog.Handler` (modify record), but it makes the conversion code more complex because `logr.Logger.level` needs to be restored when going back from such a `slog.Handler` to a `logr.Logger` (`FromSlog(ToSlog(logger.V(4))` == `logger.V(4)`).